src/directory/mod.rs

```diff
@@ -24,18 +24,29 @@ pub use self::ram_directory::RAMDirectory;
 pub use self::read_only_source::ReadOnlySource;
 pub(crate) use self::watch_event_router::WatchCallbackList;
 pub use self::watch_event_router::{WatchCallback, WatchHandle};
-use std::io::{BufWriter, Write};
+use std::io::{self, BufWriter, Write};
 
 #[cfg(feature = "mmap")]
 pub use self::mmap_directory::MmapDirectory;
 
 pub use self::managed_directory::ManagedDirectory;
 
+
+/// Trait used to indicate when no more write need to be done on a writer
+pub trait TerminatingWrite: Write {
+    /// Indicate that the writer will no longer be used
+    fn terminate(mut self) -> io::Result<()> where Self: Sized {
+        self.flush()
+    }
+}
+
+impl<W: TerminatingWrite + ?Sized> TerminatingWrite for Box<W> {}
```

Comment:
afterthought: this implementation is wrong, it won't call inner's terminate, but the default one defined just above.
It can't call inner's implementation because of `?Sized` (require otherwise Box is hardly usable at all), but removing the need for Sized on `terminate` mean it can't take by value, and the borrow check can't help us maintain the contract that a TerminatingWrite can't be used after a call to terminate